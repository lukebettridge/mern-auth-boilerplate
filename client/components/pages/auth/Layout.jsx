import React from "react";
import PropType from "prop-types";

import { Container, Heading, Paragraph } from "../../styles";

const Layout = props => {
	return (
		<Container small>
			<Heading>Webapp</Heading>
			{props.children}
			<Paragraph center light mt="l">
				Developed by{" "}
				<a
					href="https://paddl.co.uk/"
					rel="noopener noreferrer"
					target="_blank"
				>
					Paddl
				</a>
				<br />. A limited company registered in England and Wales. Registered
				number: 11992253.
			</Paragraph>
		</Container>
	);
};

Layout.propTypes = {
	children: PropType.any
};

export default Layout;
