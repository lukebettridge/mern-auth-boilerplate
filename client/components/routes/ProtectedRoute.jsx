import React from "react";
import PropType from "prop-types";
import { Route, Redirect } from "react-router-dom";
import Cookies from "js-cookie";

const ProtectedRoute = ({ component: Component, ...rest }) => {
	const isAuthenticated = () => {
		const authenticated = Cookies.get("authenticated");
		return authenticated && authenticated === "true";
	};

	return (
		<Route
			{...rest}
			render={props =>
				isAuthenticated() ? (
					<Component {...props} />
				) : (
					<Redirect
						to={{
							pathname: "/login",
							state: { from: props.location }
						}}
					/>
				)
			}
		/>
	);
};

ProtectedRoute.propTypes = {
	component: PropType.any,
	location: PropType.string
};

export default ProtectedRoute;
