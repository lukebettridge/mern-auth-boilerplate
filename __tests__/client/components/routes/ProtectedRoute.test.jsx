import React from "react";
import { mount } from "enzyme";
import toJSON from "enzyme-to-json";
import cloneDeep from "lodash/cloneDeep";
import { MemoryRouter } from "react-router-dom";
import { MockedProvider } from "@apollo/react-testing";

import { actAwait, updateWrapper } from "tests/test-utils";

import Layout from "components/pages/layout";

import Context from "components/context";
import ProtectedRoute, {
	CURRENT_USER_QUERY
} from "components/routes/ProtectedRoute";

const currentUser = {
	id: "1",
	forename: "John",
	surname: "Doe",
	email: "hello@example.com",
	roles: ["admin"],
	active: true
};
const defaultMocks = [
	{
		request: {
			query: CURRENT_USER_QUERY
		},
		result: {
			data: {
				currentUser
			}
		}
	}
];
const MockComponent = () => <div />;
let wrapper;
let subject;

describe("ProtectedRoute component", () => {
	beforeEach(() => {
		jest.resetAllMocks();
	});

	afterEach(() => {
		wrapper.unmount();
	});

	it("snapshot renders", async () => {
		mountWrapper();

		await updateSubject();
		subject = wrapper.find(ProtectedRoute);

		expect(toJSON(subject)).toMatchSnapshot();
	});

	it("renders component", async () => {
		mountWrapper();

		await updateSubject();
		subject = wrapper.find(ProtectedRoute);

		expect(subject.find(MockComponent).length).toEqual(1);
		expect(subject.find(MockComponent).props().currentUser).toEqual(
			currentUser
		);
		expect(subject.find(MockComponent).props().refetchCurrentUser).toEqual(
			expect.any(Function)
		);
	});

	it("renders loading", async () => {
		const mocks = cloneDeep(defaultMocks);
		Object.assign(mocks[0], { result: null });

		mountWrapper(mocks);

		await actAwait();

		expect(subject.find(Layout).length).toEqual(1);
		expect(subject.find(MockComponent).length).toEqual(0);
	});

	it("does not render component on error", async () => {
		const mocks = cloneDeep(defaultMocks);
		Object.assign(mocks[0], { result: null, error: new Error() });

		mountWrapper(mocks);

		await actAwait();

		expect(subject.find(MockComponent).length).toEqual(0);
	});

	it("does not render component with no data", async () => {
		const mocks = cloneDeep(defaultMocks);
		Object.assign(mocks[0].result.data, { currentUser: null });

		mountWrapper(mocks);

		await actAwait();

		expect(subject.find(MockComponent).length).toEqual(0);
	});

	it("does not render component without correct roles", async () => {
		const mocks = cloneDeep(defaultMocks);
		Object.assign(mocks[0].result.data.currentUser, { roles: [] });

		mountWrapper(mocks);

		await actAwait();

		expect(subject.find(MockComponent).length).toEqual(0);
	});
});

const mountWrapper = (mocks = defaultMocks) => {
	const context = {
		notification: { success: jest.fn() }
	};

	wrapper = mount(
		<MockedProvider addTypename={false} mocks={mocks}>
			<MemoryRouter initialEntries={[{ pathname: "/path", key: "key" }]}>
				<Context.Provider value={context}>
					<ProtectedRoute
						component={MockComponent}
						path="/path"
						roles={["admin"]}
					/>
				</Context.Provider>
			</MemoryRouter>
		</MockedProvider>
	);
	subject = wrapper.find(ProtectedRoute);
};

const updateSubject = async (amount = 0) => {
	await updateWrapper(wrapper, amount);
	subject = wrapper.find(ProtectedRoute);
};
