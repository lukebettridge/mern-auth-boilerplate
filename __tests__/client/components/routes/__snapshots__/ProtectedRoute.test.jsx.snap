// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProtectedRoute component snapshot renders 1`] = `
<ProtectedRoute
  component={[Function]}
  path="/path"
  roles={
    Array [
      "admin",
    ]
  }
>
  <Route
    path="/path"
    render={[Function]}
  >
    <Query
      fetchPolicy="no-cache"
      query={
        Object {
          "definitions": Array [
            Object {
              "directives": Array [],
              "kind": "OperationDefinition",
              "name": undefined,
              "operation": "query",
              "selectionSet": Object {
                "kind": "SelectionSet",
                "selections": Array [
                  Object {
                    "alias": undefined,
                    "arguments": Array [],
                    "directives": Array [],
                    "kind": "Field",
                    "name": Object {
                      "kind": "Name",
                      "value": "currentUser",
                    },
                    "selectionSet": Object {
                      "kind": "SelectionSet",
                      "selections": Array [
                        Object {
                          "alias": undefined,
                          "arguments": Array [],
                          "directives": Array [],
                          "kind": "Field",
                          "name": Object {
                            "kind": "Name",
                            "value": "id",
                          },
                          "selectionSet": undefined,
                        },
                        Object {
                          "alias": undefined,
                          "arguments": Array [],
                          "directives": Array [],
                          "kind": "Field",
                          "name": Object {
                            "kind": "Name",
                            "value": "forename",
                          },
                          "selectionSet": undefined,
                        },
                        Object {
                          "alias": undefined,
                          "arguments": Array [],
                          "directives": Array [],
                          "kind": "Field",
                          "name": Object {
                            "kind": "Name",
                            "value": "surname",
                          },
                          "selectionSet": undefined,
                        },
                        Object {
                          "alias": undefined,
                          "arguments": Array [],
                          "directives": Array [],
                          "kind": "Field",
                          "name": Object {
                            "kind": "Name",
                            "value": "email",
                          },
                          "selectionSet": undefined,
                        },
                        Object {
                          "alias": undefined,
                          "arguments": Array [],
                          "directives": Array [],
                          "kind": "Field",
                          "name": Object {
                            "kind": "Name",
                            "value": "roles",
                          },
                          "selectionSet": undefined,
                        },
                        Object {
                          "alias": undefined,
                          "arguments": Array [],
                          "directives": Array [],
                          "kind": "Field",
                          "name": Object {
                            "kind": "Name",
                            "value": "active",
                          },
                          "selectionSet": undefined,
                        },
                      ],
                    },
                  },
                ],
              },
              "variableDefinitions": Array [],
            },
          ],
          "kind": "Document",
          "loc": Object {
            "end": 84,
            "start": 0,
          },
        }
      }
    >
      <MockComponent
        currentUser={
          Object {
            "active": true,
            "email": "hello@example.com",
            "forename": "John",
            "id": "1",
            "roles": Array [
              "admin",
            ],
            "surname": "Doe",
          }
        }
        history={
          Object {
            "action": "POP",
            "block": [Function],
            "canGo": [Function],
            "createHref": [Function],
            "entries": Array [
              Object {
                "hash": "",
                "key": "key",
                "pathname": "/path",
                "search": "",
              },
            ],
            "go": [Function],
            "goBack": [Function],
            "goForward": [Function],
            "index": 0,
            "length": 1,
            "listen": [Function],
            "location": Object {
              "hash": "",
              "key": "key",
              "pathname": "/path",
              "search": "",
            },
            "push": [Function],
            "replace": [Function],
          }
        }
        location={
          Object {
            "hash": "",
            "key": "key",
            "pathname": "/path",
            "search": "",
          }
        }
        match={
          Object {
            "isExact": true,
            "params": Object {},
            "path": "/path",
            "url": "/path",
          }
        }
        refetchCurrentUser={[Function]}
      >
        <div />
      </MockComponent>
    </Query>
  </Route>
</ProtectedRoute>
`;
